libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+// Created by Wenhao Qu
+//
+// Maintainers:
+// Michael Wang <yun.wang@linux.alibaba.com>
+// Wenhao Qu <quxi.qwh@alibaba-inc.com>
+#include <argp.h>
+#include <stdio.h>
+#include <signal.h>
+#include <sched.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <locale.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <sys/wait.h>
+#include <dirent.h>
+#include <bpf/bpf.h>
+#include "errno_helpers.h"
+#include "trace_helpers.h"
+#include "syscall_helpers.h"
+#include "schedsnoop.h"
+#include "schedsnoop.skel.h"
+
+#define NS_IN_SEC		1000000000LLU
+#define NS_IN_MS		1000000LLU
+#define NS_IN_US		1000LLU
+#define PERF_BUFFER_PAGES	64
+
+const char *argp_program_version = "schedsnoop 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"Trace the related schedule events of a specified task.\n"
+"\n"
+"USAGE: schedsnoop -p PID [-s]\n"
+"\n"
+"EXAMPLES:\n"
+"    schedsnoop -p 49870       	# trace pid 49870\n"
+"    schedsnoop -p 49870 -s    	# trace pid 49870 and system call\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process PID to trace"},
+	{ "syscall", 's', NULL, 0, "Trace SYSCALL Info"},
+	{},
+};
+
+static struct env {
+	int nr_cpus;
+	int trace_syscall;
+	int target;
+} env;
+
+int ti_map_fd;
+int si_map_fd;
+int start_map_fd;
+int end_map_fd;
+bool exiting = false;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid;
+	switch (key) {
+	case 'p':
+		if(arg == NULL) {
+			fprintf(stderr, "PID is required\n");
+			argp_usage(state);
+		} else {
+			pid = strtol(arg, NULL, 10);
+			if (pid <= 0) {
+				fprintf(stderr, "Invalid PID: %s\n", arg);
+				argp_usage(state);
+			}
+			env.target = pid;
+		}
+		break;
+	case 's':
+		env.trace_syscall = 1;
+		break;
+	case ARGP_KEY_END:
+		if (!env.target) {
+			fprintf(stderr, "No target PID\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static inline int time_to_str(u64 ns, char *buf, size_t len)
+{
+
+	if (ns > 10 * NS_IN_SEC)
+		snprintf(buf, len, "%llus", ns / NS_IN_SEC);
+	else if (ns > 10 * NS_IN_MS)
+		snprintf(buf, len, "%llums", ns / NS_IN_MS);
+	else if (ns > 10 * NS_IN_US)
+		snprintf(buf, len, "%lluus", ns / NS_IN_US);
+	else
+		snprintf(buf, len, "%lluns", ns);
+
+	return 0;
+}
+
+static inline void pr_ti(struct trace_info *ti, char *opt, char *delay)
+{
+	char buf[27];
+	snprintf(buf, sizeof(buf), "%lluus", ti->ts / NS_IN_US);
+	printf("%-27sCPU=%-7dPID=%-7dCOMM=%-20s%-37s%-17s\n",
+				buf, ti->cpu, ti->pid, ti->comm, opt,
+				delay ? delay : "");
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct trace_info *ti = data;
+	char d_str[16];
+	char comm_buf[2 * TASK_COMM_LEN];
+	char func[80];
+	struct si_key sik;
+	u64 siv;
+	static u64 w_start, p_start, last_time;

```

Comment:
first value of last_time will be zero, which will give you huge delta for the very first sample, that's probably not what you want. You probably want to fetch kernel timestamp right before you attach BPF program.

It's not immediately obvious to me how to get same timestamp from user-space. I suspect it might be one of clock_gettime() with  CLOCK_MONOTONIC or CLOCK_MONOTONIC_RAW, but I haven't checked. It would be good to figure it out, though, I think this question comes up from time to time.

Alternatively, there is bpf_get_ktime_boot_ns(), with more well-defined origin.

Reply:
So it seems like you should be able to get the same-origin clock in user-space with clock_gettime(CLOCK_MONOTONIC), according to https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L694

Reply:
last_time is used to calculate the time duration between two specific events, which shows how long a certain tasks has been running or how long it has been preempted. So it doesn't matter that the first value is zero or the timestamp that I attach the BPF program.

The only problem is, if the target task has been running for a while before we start tracing it, the first sample may output something strange, e.g. the target task may be preempted for a very long time because of the zero first value. This seems unsolvable since we cannot get any information before we attach the BPF program. (Maybe we can read procfs to get some information? But it's hard to get any information about preemption i think.)

the further output information won't be affected.